# Exercícios: Tratamento dos Dados

#### Carga da Base de Dados Indicadores Criminais RS

In [1]:
import pandas as pd
from pandas import DataFrame

arquivo = './data/df_crimesRS.json'

df_crimes = pd.read_json(arquivo)
df_crimes.head()

index  Municípios  Homicídio  Doloso  Total de vítimas de Homicidio Doloso  \
0      0      ACEGUA                  1                                     1   
1      1  AGUA SANTA                  0                                     0   
2      2       AGUDO                  2                                     2   
3      3   AJURICABA                  1                                     1   
4      4     ALECRIM                  1                                     1   

   Latrocínio  Furtos  Abigeato*  Furto de Veículo  Roubos  Roubo de Veículo  \
0           0      65         18                 1       5                 1   
1           0      31          3                 3       0                 2   
2           0     123         11                 1       5                 0   
3           0      58          5                 2       0                 0   
4           0      85         19                 1       3                 0   

   Estelionato  Delitos Relacionados à Armas e Munições  \
0            4                                        2   
1            2                                        1   
2           19                                       13   
3            8                                        8   
4            8                                       15   

   Entorpecentes - Posse  Entorpecentes - Tráfico  Vítimas de Latrocínio  \
0                      0                        0                      0   
1                      0                        1                      0   
2                     14                       15                      0   
3                      3                        1                      0   
4                      7                        9                      0   

   Vítimas de Lesão Corp. Seg. Morte   ano  
0                                  0  2018  
1                                  0  2018  
2                                  0  2018  
3                                  0  2018  
4                                  0  2018

#### Pivoteamento Longo/Largo e Cálculo de Totais

A base de dados possui a quantidade de crimes registrados para cada ano e município do RS.<br>
Crie um novo DataFrame contendo somente os valores do crime "Homicídio  Doloso" para todos os municípios e anos.

Modifique o DataFrame de forma que o nome do município seja o índice das linhas e cada ano seja uma coluna, como abaixo:<br>
<pre>ano	        2012	2013	2014	2015	2016	2017	2018	2019	2020	2021	2022
Municípios
Nome
</pre>

Acrescente uma coluna denominada "Total/Município" e calcule o total de homicídios para cada município.

Acrescente uma linha  denominada "Total/Ano" e calcule o total de homicídios por ano para todos os municípios.

Liste as linhas contendo dados dos 10 municípios com maior quantidade de homicídios no período de 2012 a 2022, seguida da linha "Total/Ano".

Calcule o valor percentual de cada um dos municípios em relação ao total, utilizando a linha "Total/Ano" como base.

#### Pivoteamento Largo/Longo - Voltando ao Formato Original

Utilize o DataFrame em formato largo, em que foi acrescentada uma coluna e linha de totais para o exercício a seguir.<br>
O objetivo é retornar os dados ao formato original.

Primeiramente, remova do DataFrame a Coluna e a linha de Total.

Em seguida, utilize a função melt() para transformar do formato largo:
<pre>ano	        2012	2013	2014	2015	2016	2017	2018	2019	2020	2021	2022
Municípios
Nome
</pre>
Para o formato longo:
<pre>
Municípios      Ano     Homicídio Doloso
Nome            2012    123
Nome            2013    150
Nome            2013    160
</pre>
Se necessário, consulte a documentação do método pd.melt() em https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.melt.html<br>
Lembre-se de designar a coluna "Municípios" como índice do DataFrame resultante.

#### Carga Base de Dados: População Municípios 2010/2020

In [ ]:
df_populacao = pd.read_csv('./data/populacao_municipios.csv',sep=';')
df_populacao.head()

Vamos calcular a taxa de homicídios por habitante para cada um dos municípios do Rio Grande do Sul.<br>
Para isso, será necessário juntar os DataFrames df_populacao e df_homicidio utilizando o campo "Município" como chave.

#### Preparação dos Dados

Vez que o campo Código não está disponível na base de homicídios ,teremos que fazer uma junção utilizando o nome do município como chave.<br>
Note que, na primeira base, o nome dos municípios está em caixa alta e sem acentos, como "AGUA SANTA".<br>
Na segunda base, os nomes estão grafados da forma usual, como "Água Santa'.<br>
Utilize a seguinte função auxiliar limpa_texto() para modificar os valores da coluna Município do DataFrame df_populacao:

In [ ]:
import unicodedata

def limpa_texto(texto='',*kword):
    nfkd_form = unicodedata.normalize('NFKD', texto)
    only_ascii = nfkd_form.encode('ASCII', 'ignore').decode("utf-8")
    texto = only_ascii.upper()
    return texto

In [ ]:
df_populacao['Município']=df_populacao['Município'].apply(limpa_texto)
df_populacao

A base de dados contém dados de todos os estados, o código do município e a população nos anos de 2010 e 2022.<br>
Crie um novo DataFrame contendo somente os dados do Estado do RS, código e a população do ano de 2022.<br>
Modifique o nome das colunas de '2022' para 'População' e "Município' para 'Municípios'.<br>
Utilize o nome do município, como índice das linhas.<br>

Faça a junção dos DataFrames de população de homicídios e população dos municípios do RS por meio do índice.<br>
Se necessário consulte a documentação do método pd.merge() disponível em: https://pandas.pydata.org/docs/reference/api/pandas.merge.html<br>
Verifique os parâmetros lef_index e right_index.

Armazene os dados do DataFrame em um arquivo JSON para utilização nas próximas aulas.